### ISSUE TITLE
wxAuiNotebook: Adding a page with the selected flag does not change selection 

### ISSUE THREAD

**Issue by robertop** [nhold/wxWidgets#76]
This issue was reported in the wx-dev mailing list, creating a github issue for the sake of completeness

When adding a page to a wxAuiNotebook with the select argument being true, the notebook does not switch selection.

Steps to reproduce:
1. apply the patch below
2. run the auidemo
3. go to the  menu view .. Create wxAuiNotebook
4. go to the menu view ..  Add page to notebook
5. observe that the page is added, but the notebook does not switch the selected tab as it was instructed to.

Patch 

```
diff --git a/samples/aui/auidemo.cpp b/samples/aui/auidemo.cpp
index 12d7776..c5e384f 100644
--- a/samples/aui/auidemo.cpp
+++ b/samples/aui/auidemo.cpp
@@ -103,6 +103,8 @@ class MyFrame : public wxFrame
         ID_DropDownToolbarItem,
         ID_SampleItem,
         ID_ToggleNB,
+       ID_AddContentNotebookPage,
+       ID_ContentNotebook,
         ID_FirstPerspective = ID_CreatePerspective+1000
     };

@@ -158,6 +160,7 @@ private:
     void OnUpdateUI(wxUpdateUIEvent& evt);
     void OnToggleNB(wxCommandEvent& WXUNUSED(evt));
     void OnPaneClose(wxAuiManagerEvent& evt);
+   void OnAddContentNotebookPage(wxCommandEvent& event);

 private:

@@ -615,6 +618,7 @@ BEGIN_EVENT_TABLE(MyFrame, wxFrame)
     EVT_MENU(wxID_EXIT, MyFrame::OnExit)
     EVT_MENU(wxID_ABOUT, MyFrame::OnAbout)
     EVT_MENU(ID_ToggleNB, MyFrame::OnToggleNB)
+   EVT_MENU(ID_AddContentNotebookPage, MyFrame::OnAddContentNotebookPage)
     EVT_UPDATE_UI(ID_AllowFloating, MyFrame::OnUpdateUI)
     EVT_UPDATE_UI(ID_TransparentHint, MyFrame::OnUpdateUI)
     EVT_UPDATE_UI(ID_VenetianBlindsHint, MyFrame::OnUpdateUI)
@@ -667,6 +671,7 @@ MyFrame::MyFrame(wxWindow* parent,
     view_menu->Append(ID_HideActivePane, _("Hide active pane"));
     view_menu->Append(ID_FloatActivePane, _("Float active pane"));
     view_menu->Append(ID_DockActivePane, _("Dock active pane"));
+   view_menu->Append(ID_AddContentNotebookPage, _("Add page to notebook"));


     wxMenu* manager_options_menu = new wxMenu;
@@ -1287,7 +1292,7 @@ wxAuiNotebook* MyFrame::CreateNotebook()
    // create the notebook off-window to avoid flicker
    wxSize client_size = GetClientSize();

-   wxAuiNotebook* ctrl = new wxAuiNotebook(this, wxID_ANY,
+   wxAuiNotebook* ctrl = new wxAuiNotebook(this, ID_ContentNotebook,
                                     wxPoint(client_size.x, client_size.y),
                                     wxSize(430,200),
                                     wxAUI_NB_DEFAULT_STYLE | wxAUI_NB_TAB_EXTERNAL_MOVE | wxNO_BORDER);
@@ -1365,3 +1370,23 @@ wxString MyFrame::GetIntroText()
         "</body></html>";
     return wxString::FromAscii(text);
 }
+
+void MyFrame::OnAddContentNotebookPage(wxCommandEvent& WXUNUSED(event))
+{
+   wxWindow* win = this->FindWindowById(ID_ContentNotebook, this);
+   if (!win) {
+       wxMessageBox(_("No notebook has been created"));
+       return;
+   }
+   wxAuiNotebook* notebook = wxDynamicCast(win, wxAuiNotebook);
+   if (!notebook) {
+       wxMessageBox(_("No notebook has been created"));
+       return;
+   }
+   
+   wxTextCtrl* txt = new wxTextCtrl( notebook, wxID_ANY, wxT("Newly created tab"), 
+       wxDefaultPosition, wxDefaultSize, wxTE_MULTILINE|wxNO_BORDER);
+   notebook->AddPage( txt, wxT("Newly created tab"), true );
+   
+   m_mgr.Update();
+}
\ No newline at end of file
```

**Comment by Kinaou** [comment]
I have a similar issue but just by adding (programmatically) several panes at the same place (consequently a notebook is created).
1. I tried `wxAuiManager::AddPane` followed by `SetActivePane` but the newly added pane/tab is not selected because at the call to `SetActivePane` the page is not yet added to the notebook.
   And if I call `Update` between the two commands, the first tab then the last tab are selected alternatively, which is annoying visually (but the good tab is selected).
2. I tried to define both the flags `optionActive` and `optionActiveNotebook` to my `wxAuiPaneInfo` but at the add of the second pane, the notebook is tiled horizontally, all the tabs being on the same tile.
   This is because several panes have the `optionActiveNotebook` flag. The issue is potentially known because in the method `wxAuiManager::LayoutAddDock` there is a comment `// Add a debug warning` when the case is encountered (see commit https://github.com/Kinaou/wxWidgets/commit/1dffc8a2df6d72383824c8ba99e30221dafca675)

**Comment by dronesalot** [comment]
I think the following problem is a side effect of this:
Create a pane with a splitter then try to drag one of the two tabs. 
I get framemanager.cpp(2719) assert failure in layoutadddock flag optionactivenotebook should not be defined to several wxaudpaneinfo. I think with a splitter in there it is valid to have 2 active at the same time.

**Comment by Kinaou** [comment]
You are right, it does not work very well yet.
By cons, I'm not sure to understand perfectly the need of having the two flags `optionActive` and `optionActiveNotebook`. Could the author enlighten me on the subject?

**Comment by mjmacleod** [comment]
optionActive indicates that the pane is the currently active pane (AUI can be set to have an 'active' pane - only one pane is ever active at a time)

optionActiveNotebook indicates that the pane is the currently visible pane of a notebook (not the 'physical' notebook class e.g. wxAuiNotebook but the internal 'logical' notebooks that AUI creates on the fly), the one whose tab is selected.

Within an AUI manager only one pane should ever be 'optionActive' at a time, while you may have multiple 'optionActiveNotebook' panes if there are multiple dynamic notebooks being created.

A pane can therefore be both 'optionActive' and 'optionActiveNotebook' at the same time (if you have just selected it), or it can be just 'optionActiveNotebook' without being 'optionActive' (if you have just given the selection to another pane that isn't part of the same dynamic notebook) but it can never be just 'optionActive' without also being 'optionActveNotebook' unless it is not currently part of a notebook at all.
I hope this makes sense?

**Comment by Kinaou** [comment]
It is quite clear, thank you for the explanations.
I will address the @fastworks issue as soon as I have a moment.

**Comment by dronesalot** [comment]
I was giving some thought to this issue. I think wxAuiPaneInfo needs a pointer to its active child. 

Making a different child (tab) active consists of adjusting this pointer. Deleting the active tab means picking a new active one or setting it to NULL if there are no more children. It doesn't make sense to have to traverse all the child objects to find one flagged as being active. This means setting the active tab is called at the book level not the tab level. This avoids all the extra checks and balances to make sure there are not 2 active tabs at the same time. It also makes dropping a tab on a new book trivial because you're just updating it's active pane.

**Comment by mjmacleod** [comment]
Sorry I really don't see what you mean here by 'child' or what benefit it would bring.

**Comment by mjmacleod** [comment]
It also wouldn't solve any issues as it would still need to check for each logical notebook whether the 'active' tab is still part of it or has been moved, further pointers can't be serialised across runs so save/load perspective would still need to work with options.

**Comment by mjmacleod** [comment]
If you can create testcases for both of the issues reported in this thread I can have a quick look for a proper solution - I think the current proposed solutions both gloss over a lot of internal edge cases that will be problematic.

**Comment by dronesalot** [comment]
Here is what I'm referring to:
framemanager.cpp: 629

```
int i, paneCount;
wxAuiPaneInfo* activePaneInfo = NULL;
for (i = 0, paneCount = m_panes.GetCount(); i < paneCount; ++i)
{
    wxAuiPaneInfo& pane = m_panes.Item(i);
    pane.SetFlag(wxAuiPaneInfo::optionActive, false);
    if (pane.GetWindow() == activePane)
    {
        activePaneInfo = &pane;
        wxAuiTabContainer* ctrl;
        int ctrlIndex;
        if(FindTab(pane.GetWindow(), &ctrl, &ctrlIndex))
        {
            ctrl->SetActivePage(pane.GetWindow());
        }

        pane.SetFlag(wxAuiPaneInfo::optionActive, true);
    }
}
```

We traverse all the panes trying to see which one has the active flag set. Since it's already specified that there can be only 1 active pane then why does the auimanager not carry a pointer or reference to the active pane? 

In many cases I'll have 50 pages open at one time. AUI becomes noticeably laggy on slower tablets. I'm just hoping that something could make it more efficient.

**Comment by mjmacleod** [comment]
While I appreciate the performance concerns, and I'm quite confident there are several ways to fix those, can we stick to one issue at a time :)
I suggest that performance enhancements are something that we would be better able to deal with after a merge.

I'll try land a fix for the issue at hand later.

**Comment by mjmacleod** [comment]
```
diff --git a/src/aui/auibook.cpp b/src/aui/auibook.cpp
index 7bcde4b..da10925 100644
--- a/src/aui/auibook.cpp
+++ b/src/aui/auibook.cpp
@@ -230,10 +230,16 @@ bool wxAuiNotebook::InsertPage(size_t pageIndex, wxWindow* page, const wxString&
     wxAuiDoInsertPage(m_mgr.GetAllPanes(),1,0,1,0,pageIndex);

     m_mgr.AddPane(page, wxAuiPaneInfo().Centre().Layer(0).Position(0).Caption(caption).Floatable(false).Movable().Page(pageIndex).Icon(bitmap).Dockable(m_mgr.HasFlag(wxAUI_NB_TAB_SPLIT)).CloseButton(false).AlwaysDockInNotebook());
-    
-    // Change the selection if explicitly requested, or if the page is the first one; for the later case, it ensures that there is a current page in the notebook   
+
+    // Change the selection if explicitly requested, or if the page is the first one; for the later case, it ensures that there is a current page in the notebook.
     if (select || GetPageCount() == 1)
+    {
+        // Note - the double 'Update()' call here is intentional, unfortunately due to the way things work, the pane only actually becomes part of a notebook after Update calculates everything.
+        // So without the 'Update()' call 'SetSelection()' has no effect.
+        // An alternative here that we could explore is to instead change 'SetSelection' to work more like 'wxAuiDoInsertPage()'
+        m_mgr.Update();
         SetSelection(pageIndex);
+    }

     m_mgr.Update();
```

Should do it, however there seems to be a second issue where the selected tab does not scroll into view if it is off screen - the code to do this executes but it looks like the scrolling is later reset for some reason. Will look into that first before doing a final pull request.

**Comment by mjmacleod** [comment]
I've narrowed down the scrolling problem to this piece of code.

```
-    while (m_tabOffset > 0 && IsTabVisible(pageCount-1, m_tabOffset-1, &dc, wnd) )
-            --m_tabOffset;
+    //while (m_tabOffset > 0 && IsTabVisible(pageCount-1, m_tabOffset-1, &dc, wnd) )
+            //--m_tabOffset;
```

If I apply this patch it then works as expected. This piece of code looks suspect to me as it permanently alters the member variable? However I'm struggling to make 100% sense of this code, perhaps I'm just to tired.
Can anyone make any sense of what is going on there and any ideas on what the code should actually do instead?

**Comment by mjmacleod** [comment]
Anyway, above is my suggested solution to this. Any input from anyone who understands what is going on appreciated :)

**Comment by pkulchenko** [comment]
@nhold, was this issue closed by mistake? I don't see @mjmacleod's suggestions implemented and the problem with multiple selected/active tabs is still present (per #119, #96).

**Comment by pkulchenko** [comment]
@nhold, never mind; I see that it's been resolved in #85 and #119 describes a slightly different issue.
